### PR TITLE
feat: adds configurable token lifetime support

### PR DESCRIPTION
**Pull request by aeitzman**
Adds support for configurable token lifespan via the service_account_impersonation object in the ADC. Also adds unit + integration tests.
See go/byoid-sdk-service-account-impersonation-ctl

### CODE REVIEW

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 1]
```diff
@@ -73,8 +75,46 @@ abstract static class CredentialSource {
...
+   *     "token_lifetime_seconds": 2800
+   *    }
+   * }
+   * </pre>
+   */
+  static class ServiceAccountImpersonationOptions {
+    public final int lifetime;
+
+    ServiceAccountImpersonationOptions(Map<String, Object> serviceAccountImpersonationOptionsMap) {
+      if (serviceAccountImpersonationOptionsMap.containsKey(TOKEN_LIFETIME_SECONDS_KEY)) {
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:98]: we can reduce nesting here:

```
if (!options...) {
  lifetime = ..
  return;
} 
...

see go/tott/553

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 2]
```diff
@@ -73,8 +75,46 @@ abstract static class CredentialSource {
...
+   *   "service_account_impersonation": {
+   *     "token_lifetime_seconds": 2800
+   *    }
+   * }
+   * </pre>
+   */
+  static class ServiceAccountImpersonationOptions {
+    public final int lifetime;
+
+    ServiceAccountImpersonationOptions(Map<String, Object> serviceAccountImpersonationOptionsMap) {
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:97]: nit: `serviceAccountImpersonationOptionsMap` seems redundant - how about `optionsMap`?

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 3]
```diff
@@ -73,8 +75,46 @@ abstract static class CredentialSource {
...
   }
 
+  /**
+   * Encapsulates the service account impersonation options portion of the configuration for
+   * ExternalAccountCredentials.
+   *
+   * <p>If token_lifetime_seconds is not specified, the library will default to a 1-hour lifetime.
+   *
+   * <pre>
+   * Sample credential source for Pluggable Auth credentials:
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:85]: This actually isn't in the credential source. Also should just be generic to External Credentials

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 4]
```diff
@@ -73,8 +75,46 @@ abstract static class CredentialSource {
...
+      } else {
+        lifetime = DEFAULT_TOKEN_LIFETIME_SECONDS;
+      }
+    }
+  }
+
   private static final String CLOUD_PLATFORM_SCOPE =
       "https://www.googleapis.com/auth/cloud-platform";
+  private static final int DEFAULT_TOKEN_LIFETIME_SECONDS = 3600;
+  private static final String TOKEN_LIFETIME_SECONDS_KEY = "token_lifetime_seconds";
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:117]: Should be inside ServiceAccountImpersonationOptions

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 5]
```diff
@@ -175,6 +217,44 @@ protected ExternalAccountCredentials(
...
+
+  /**
+   * See {@link ExternalAccountCredentials#ExternalAccountCredentials(HttpTransportFactory, String,
+   * String, String, CredentialSource, String, String, String, String, String, Collection,
+   * EnvironmentProvider)}
+   *
+   * @param serviceAccountImpersonationOptions additional options for service account impersonation,
+   *     may be null.
+   */
+  protected ExternalAccountCredentials(
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:244]: Don't need to add a new constructor here

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 6]
```diff
@@ -738,6 +844,12 @@ Builder setEnvironmentProvider(EnvironmentProvider environmentProvider) {
       return this;
     }
 
+    Builder setServiceAccountImpersonationOptions(
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:847]: Should add some documentation

File: oauth2_http/javatests/com/google/auth/oauth2/ITWorkloadIdentityFederationTest.java [hunk 7]
```diff
@@ -168,6 +170,29 @@ public void pluggableAuthCredentials() throws IOException {
     callGcs(pluggableAuthCredentials);
   }
 
+  /**
+   * Sets the service account impersonation object in configuration JSON with a non-default value
+   * for token_lifetime_seconds and validates that the lifetime is used for the access token.
+   */
+  @Test
+  public void IdentityPoolCredentialsWithServiceAccountImpersonationOptions() throws IOException {
```
Comment (lsirac) [inline oauth2_http/javatests/com/google/auth/oauth2/ITWorkloadIdentityFederationTest.java:178]: identityPoolCredentials_withServiceAccountImpersonationOptions

File: oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountCredentialsTest.java [hunk 8]
```diff
@@ -202,6 +202,27 @@ public void fromJson_identityPoolCredentialsWorkforce() {
     assertNotNull(credential.getCredentialSource());
   }
 
+  @Test
```
Comment (lsirac) [inline oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountCredentialsTest.java:205]: Missing refreshAccessToken tests for all of these credentials (in their respective files, not this one)

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 9]
```diff
@@ -73,6 +75,45 @@ abstract static class CredentialSource {
...
+   *     "token_lifetime_seconds": 2800
+   *    }
+   * }
+   * </pre>
+   */
+  static class ServiceAccountImpersonationOptions {
+    private static final int DEFAULT_TOKEN_LIFETIME_SECONDS = 3600;
+    private static final String TOKEN_LIFETIME_SECONDS_KEY = "token_lifetime_seconds";
+
+    public final int lifetime;
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:98]: private & add a package private getter

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 10]
```diff
@@ -73,6 +75,45 @@ abstract static class CredentialSource {
...
+   * Sample configuration:
+   * {
+   *   ...
+   *   "service_account_impersonation": {
+   *     "token_lifetime_seconds": 2800
+   *    }
+   * }
+   * </pre>
+   */
+  static class ServiceAccountImpersonationOptions {
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:94]: final

File: README.md [hunk 11]
```diff
@@ -467,6 +468,34 @@ credentials unless they do not meet your specific requirements.
 You can now [use the Auth library](#using-external-identities) to call Google Cloud
 resources from an OIDC or SAML provider.
 
+#### Configurable Token Lifetime
+When creating a credential configuration with workload identity federation using service account impersonation, you can provide an optional argument to configure the access token lifetime.
```
Comment (lsirac) [inline README.md:472]: configure the service account access token lifetime

File: README.md [hunk 12]
```diff
@@ -467,6 +468,34 @@ credentials unless they do not meet your specific requirements.
...
+
+To generate the configuration with configurable token lifetime, run the following command (This example uses an AWS configuration, but the token lifetime can be configured for all workload identity federation providers):
```
Comment (lsirac) [inline README.md:474]: this

File: README.md [hunk 13]
```diff
@@ -467,6 +468,34 @@ credentials unless they do not meet your specific requirements.
...
+- `$AWS_PROVIDER_ID`: The AWS provider ID.
+- `$SERVICE_ACCOUNT_EMAIL`: The email of the service account to impersonate.
+- `$TOKEN_LIFETIME`: The desired lifetime duration of the service account access token in seconds.
+
+The `service-account-token-lifetime-seconds` flag is optional. If not provided, this defaults to one hour.
+The minimum allowed value is 600 (10 minutes) and the maximum allowed value is 43200 (12 hours).
+If a lifetime greater than one hour is required, the service account must be added as an allowed value in an Organization Policy that enforces the `constraints/iam.allowServiceAccountCredentialLifetimeExtension` constraint.
+Note that configuring a short lifetime (e.g. 10 minutes) will result in the library initiating the entire token exchange flow every 10 minutes, which will call the 3rd party token provider even if the 3rd party token is not expired.
+
+
```
Comment (lsirac) [inline README.md:497]: remove extra new lines

File: README.md [hunk 14]
```diff
@@ -467,6 +468,34 @@ credentials unless they do not meet your specific requirements.
...
+Where the following variables need to be substituted:
+- `$PROJECT_NUMBER`: The Google Cloud project number.
+- `$POOL_ID`: The workload identity pool ID.
+- `$AWS_PROVIDER_ID`: The AWS provider ID.
+- `$SERVICE_ACCOUNT_EMAIL`: The email of the service account to impersonate.
+- `$TOKEN_LIFETIME`: The desired lifetime duration of the service account access token in seconds.
+
+The `service-account-token-lifetime-seconds` flag is optional. If not provided, this defaults to one hour.
+The minimum allowed value is 600 (10 minutes) and the maximum allowed value is 43200 (12 hours).
+If a lifetime greater than one hour is required, the service account must be added as an allowed value in an Organization Policy that enforces the `constraints/iam.allowServiceAccountCredentialLifetimeExtension` constraint.
```
Comment (lsirac) [inline README.md:494]: add a new line

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 15]
```diff
@@ -738,6 +804,13 @@ Builder setEnvironmentProvider(EnvironmentProvider environmentProvider) {
       return this;
     }
 
+    /** Sets the optional service account impersonation options. */
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:807]: Should be public. Can take in a map if you don't want to expose ServiceAccountImpersonationOptions

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 16]
```diff
@@ -733,6 +803,13 @@ public Builder setWorkforcePoolUserProject(String workforcePoolUserProject) {
...
+    public Builder setServiceAccountImpersonationOptions(
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:807]: ServiceAccountImpersonationOptions is package private. Should either be public or pass a map here?

**Review by lsirac — APPROVED** [review]

File: oauth2_http/javatests/com/google/auth/oauth2/AwsCredentialsTest.java [hunk 17]
```diff
@@ -142,6 +143,41 @@ public void refreshAccessToken_withServiceAccountImpersonation() throws IOExcept
...
+            AwsCredentials.newBuilder(AWS_CREDENTIAL)
+                .setTokenUrl(transportFactory.transport.getStsUrl())
+                .setServiceAccountImpersonationUrl(
+                    transportFactory.transport.getServiceAccountImpersonationUrl())
+                .setHttpTransportFactory(transportFactory)
+                .setCredentialSource(buildAwsCredentialSource(transportFactory))
+                .setServiceAccountImpersonationOptions(
+                    new ServiceAccountImpersonationOptions(
+                        ExternalAccountCredentialsTest.buildServiceAccountImpersonationOptions(
+                            2800)))
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/AwsCredentialsTest.java:164]: I think you are missing more tests on corner cases of value parsing, string type, big int.. 

if you decide to fall back to one hour in case of error - test that as well

**Review by TimurSadykov — CHANGES_REQUESTED** [review]

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 18]
```diff
@@ -73,6 +75,49 @@ abstract static class CredentialSource {
...
+   * }
+   * </pre>
+   */
+  public static final class ServiceAccountImpersonationOptions {
+    private static final int DEFAULT_TOKEN_LIFETIME_SECONDS = 3600;
+    private static final String TOKEN_LIFETIME_SECONDS_KEY = "token_lifetime_seconds";
+
+    private final int lifetime;
+
+    ServiceAccountImpersonationOptions(Map<String, Object> optionsMap) {
```
[the remark this answers is not in the document]
Comment (aeitzman) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:100]: lifetime is just the first property in service account options, we will likely extend it in the future to include more properties which will not necessarily be int. If the optionsMap gets changed to <String, int> we would need to make a breaking change in the future to change it back to object when more properties get added.

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 19]
```diff
@@ -73,6 +75,49 @@ abstract static class CredentialSource {
...
+      if (!optionsMap.containsKey(TOKEN_LIFETIME_SECONDS_KEY)) {
+        lifetime = DEFAULT_TOKEN_LIFETIME_SECONDS;
+        return;
+      }
+
+      Object timeout = optionsMap.get(TOKEN_LIFETIME_SECONDS_KEY);
```
[the remark this answers is not in the document]
Comment (aeitzman) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:106]: Should throw here - will match the implementation for the other already launched languages (node/python/go) and we don't want to silently ignore what the user configured in the ADC.

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 20]
```diff
@@ -608,6 +631,54 @@ private static boolean isValidUrl(List<Pattern> patterns, String url) {
...
+
+    private final int lifetime;
+
+    ServiceAccountImpersonationOptions(Map<String, Object> optionsMap) {
+      if (!optionsMap.containsKey(TOKEN_LIFETIME_SECONDS_KEY)) {
+        lifetime = DEFAULT_TOKEN_LIFETIME_SECONDS;
+        return;
+      }
+
+      try {
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:662]: The doc says that the minimum allowed is 600 sec, we probably want a precondition for that? and test.

**Review by TimurSadykov — CHANGES_REQUESTED** [review]
Mostly good, noticed the minimal duration not covered.

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java (hunk 20 above)
Comment (aeitzman) [reply oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:664]: Added min/max checks in the options constructor

**Review by TimurSadykov — APPROVED** [review]
Looks good, Thanks!
